sql/src/main/java/io/crate/operation/user/UserPrivileges.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.user;
+
+import com.carrotsearch.hppc.IntObjectHashMap;
+import io.crate.analyze.user.Privilege;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+class UserPrivileges implements Iterable<Privilege> {
+
+    @Nullable
+    private static Privilege matchPrivilege(IntObjectHashMap<Privilege> privileges,
+                                            Privilege.Type type,
+                                            Privilege.Clazz clazz,
+                                            @Nullable String ident) {
+        Privilege foundPrivilege = privileges.get(Privilege.hashCode(Privilege.State.GRANT, type, clazz, ident));
```

Comment:
it is not safe to use a hashcode, those hashcodes can conflict. 
i think in order to optimize this you probably need to split out PrivilegeIdent from Privilege where PrivilegeIdent just has type, clazz and ident and use this as the key.